### ISSUE TITLE
Type inferred as `Never` from cyclic * import

### ISSUE THREAD

**Issue by MatthewMckee4** [astral-sh/ty#113]
### Summary

My reproduction is maybe not great, but there is a package `klayout` that contains a `.pyi` file with > 75,000 lines.
```bash
mkdir /tmp/test-repro
cd /tmp/test-repro
uv venv
uv pip install klayout
echo "import klayout.db as kdb

from typing_extensions import reveal_type


def _(x: kdb.LayerInfo, y: kdb.LEFDEFReaderConfiguration):
    reveal_type(x)
    reveal_type(y)
" > main.py
uv run <red-knot-bin> check main.py
```

This gives the following error:
```bash
error: lint:invalid-type-form: Variable of type `Never` is not allowed in a type expression
 --> /tmp/test-repro/main.py:6:10
  |
6 | def _(x: kdb.LayerInfo):
  |          ^^^^^^^^^^^^^
7 |     reveal_type(x)
  |

info: revealed-type: Revealed type
 --> /tmp/test-repro/main.py:7:5
  |
6 | def _(x: kdb.LayerInfo):
7 |     reveal_type(x)
  |     ^^^^^^^^^^^^^^ `Unknown`
  |

Found 2 diagnostics
```

Also to note `LayerInfo` is defined at line 36702 and `LEFDEFReaderConfiguration` at line 35628 (and `LEFDEFReaderConfiguration` works fine)

**Comment by carljm** [comment]
Here's a minimal repro:

`main.py`:
```py
from pkg.sub import A

reveal_type(A)  # reveals Never
```

`pkg/outer.py`:
```py
class A: ...
```

`pkg/sub/__init__.py`:
```py
from ..outer import *
from .inner import *
```

`pkg/sub/inner.py`:
```py
from pkg.sub import A
```

**Comment by MatthewMckee4** [comment]
Thank you!

**Comment by MatthewMckee4** [comment]
Also, my comment about the lines doesn't matter

**Comment by MatthewMckee4** [comment]
Also, this works (as in it fails) for a mdtest

## Relative import to `.pyi`

```toml
[environment]
python = "/src/.venv"
python-version = "3.13"
```

`/src/.venv/<path-to-site-packages>/package/sub/__init__.py`:

```py
from ..bar import *
from .inner import *
```

`/src/.venv/<path-to-site-packages>/package/sub/inner.py`:

```py
from package.sub import A
```

`/src/.venv/<path-to-site-packages>/package/bar.pyi`:

```pyi
class A: ...
```

`/src/main.py`:

```py
import package.sub as sub

reveal_type(sub.A)  # revealed: Literal[A]
```

**Comment by MatthewMckee4** [comment]
And, Importantly, this passes
## Relative import to `.pyi`

```toml
[environment]
python = "/src/.venv"
python-version = "3.13"
```

`/src/.venv/<path-to-site-packages>/package/sub/__init__.py`:

```py
from ..bar import *
reveal_type(A)  # revealed: Literal[A]
from .inner import *
reveal_type(A)  # revealed: Never
```

`/src/.venv/<path-to-site-packages>/package/sub/inner.py`:

```py
from package.sub import A
```

`/src/.venv/<path-to-site-packages>/package/bar.pyi`:

```pyi
class A: ...
```

`/src/main.py`:

```py
import package.sub as sub

reveal_type(sub.A)  # revealed: Never
```

**Comment by Skylion007** [comment]
This is a massive issue in torch as it prevents `from torch import Tensor` from working with ty.

**Comment by carljm** [comment]
> This is a massive issue in torch as it prevents `from torch import Tensor` from working with ty.

Although the symptom is basically identical, that's actually a different issue, and it's fixed now with the merge of https://github.com/astral-sh/ruff/pull/17856

The issue described above still repros, however; it seems less common.

**Comment by MatthewMckee4** [comment]
I think this is resolved? my repro works fine now, and on my codebase i dont have the issue anymore

**Comment by carljm** [comment]
I still see the type of `A` wrongly revealed as `Never` in the MRE I showed in https://github.com/astral-sh/ty/issues/113#issuecomment-2859014192 so I think there is still something to fix here.

**Comment by MatthewMckee4** [comment]
Ah i see, my mistake. I wonder what the difference is between our repros then.

**Comment by sharifhsn** [comment]
As of https://github.com/astral-sh/ruff/pull/20566, the inferred type is `Divergent` not `Never`, more clearly signaling that this is a recursive problem.

@MatthewMckee4's repro was fixed by https://github.com/astral-sh/ruff/pull/17856 since `klayout` uses `__all__`, which ty checks to see that `LayerInfo` isn't available and therefore doesn't try to re-export it through the cyclic path.
